plugins/serialization/src/main/java/org/qbicc/plugin/serialization/BuildtimeHeap.java

```diff
@@ -84,18 +104,37 @@ public GlobalVariableElement getAndRegisterGlobalClassArray(ExecutableElement or
         return classArrayGlobal;
     }
 
-    public synchronized Data serializeStringLiteral(String value) {
-        // String literals are interned via equals, not ==
-        if (stringLiterals.containsKey(value)) {
-            return stringLiterals.get(value);
+    public synchronized Data serializeVmObject(VmObject value) {
+        if (vmObjects.containsKey(value)) {
+            return vmObjects.get(value);
         }
-        LoadedTypeDefinition jls = ctxt.getBootstrapClassContext().findDefinedType("java/lang/String").load();
-        Data sl = serializeObject(jls, value);
-        stringLiterals.put(value, sl);
+        PhysicalObjectType ot = value.getObjectType();
+        Data sl;
+        if (ot instanceof ClassObjectType) {
+            if (value instanceof VmClass) {
+                if (value instanceof VmArrayClass) {
+                    if (((VmArrayClass)value).getInstanceObjectType().getElementType() instanceof ReferenceArrayObjectType) {
+                        ctxt.error("Serialization of reference array class objects not supported: "+((VmArrayClass)value).getInstanceObjectType());
+                    }
+                }
+                // Redirect to the class objects already serialized by the ANALYZE post hook ClassObjectSerializer.
+                sl = classObjects.get(((VmClass)value).getTypeDefinition());
+                if (sl == null) {
+                    ctxt.error("Serializing java.lang.Class instance for unreachable class: "+((VmClass)value).getTypeDefinition());
+                }
+            } else {
+                sl = serializeVmObject((ClassObjectType) ot, value);
+            }
+        } else if (ot instanceof ReferenceArrayObjectType) {
+            sl = serializeRefArray((ReferenceArrayObjectType) ot, (VmArray)value);
+        } else {
+            sl = serializePrimArray((PrimitiveArrayObjectType) ot, (VmArray)value);
+        }
+        vmObjects.put(value, sl);
         return sl;
     }
 
-    public Data serializeClassObject(Primitive primitive) {
+    public synchronized Data serializeClassObject(Primitive primitive) {
```

Comment:
It looks like this method is still not being called; was that an oversight or is there still more to do up above? Or is it being called from someplace I can't see?

Reply:
It's being called from ClassObjectSerializer (Analyze post hook).  Was already being called, so it doesn't show up in this PR.